instrumentation/servlet/servlet-3.0/javaagent-tests/src/test/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/tomcat/TomcatServlet3Test.java

```diff
@@ -0,0 +1,220 @@
+
+    File applicationDir = new File(baseDir, "/webapps/ROOT");
+    if (!applicationDir.exists()) {
+      applicationDir.mkdirs();
+      applicationDir.deleteOnExit();
+    }
+
+    Context servletContext =
+        tomcatServer.addWebapp(getContextPath(), applicationDir.getAbsolutePath());
+    // Speed up startup by disabling jar scanning:
+    servletContext.getJarScanner().setJarScanFilter((jarScanType, jarName) -> false);
+
+    setupServlets(servletContext);
+
+    ((StandardHost) tomcatServer.getHost())
+        .setErrorReportValveClass(ErrorHandlerValve.class.getName());
+    tomcatServer.getHost().getPipeline().addValve(accessLogValue);
+
+    tomcatServer.start();
+
+    return tomcatServer;
+  }
+
+  @BeforeEach
+  void setUp() {
+    accessLogValue.getLoggedIds().clear();
+    testing().clearAllExportedData();
+  }
+
+  @Override
+  public void stopServer(Tomcat server) throws LifecycleException {
+    server.stop();
+    server.destroy();
+  }
+
+  @Override
+  public void addServlet(Context servletContext, String path, Class<? extends Servlet> servlet)
+      throws Exception {
+    String name = UUID.randomUUID().toString();
+    Tomcat.addServlet(servletContext, name, servlet.getConstructor().newInstance());
+    servletContext.addServletMappingDecoded(path, name);
+  }
+
+  @ParameterizedTest
+  @CsvSource({"1", "4"})
+  void access_log_has_ids_for__count_requests(int count) {
+    setUp();
+    AggregatedHttpRequest request = request(ACCESS_LOG_SUCCESS, "GET");
+
+    IntStream.range(0, count)
+        .mapToObj(i -> client.execute(request).aggregate().join())
+        .forEach(
+            response -> {
+              assertThat(response.status().code()).isEqualTo(ACCESS_LOG_SUCCESS.getStatus());
+              assertThat(response.contentUtf8()).isEqualTo(ACCESS_LOG_SUCCESS.getBody());
+            });
+
+    accessLogValue.waitForLoggedIds(count);
+    assertThat(accessLogValue.getLoggedIds().size()).isEqualTo(count);
+    List<String> loggedTraces =
+        accessLogValue.getLoggedIds().stream().map(Map.Entry::getKey).collect(Collectors.toList());
+    List<String> loggedSpans =
+        accessLogValue.getLoggedIds().stream()
+            .map(Map.Entry::getValue)
+            .collect(Collectors.toList());
+
+    Consumer<TraceAssert> check =
+        trace ->
+            trace.hasSpansSatisfyingExactly(
+                span -> assertServerSpan(span, "GET", ACCESS_LOG_SUCCESS, SUCCESS.getStatus()),
+                span -> assertControllerSpan(span, null));
+    testing()
+        .waitAndAssertTraces(
+            IntStream.range(0, count).mapToObj(i -> check).collect(Collectors.toList()));
+
+    List<List<SpanData>> traces = TelemetryDataUtil.groupTraces(testing().getExportedSpans());
+
+    for (int i = 0; i < count; i++) {
+      assertThat(loggedTraces).contains(traces.get(i).get(0).getTraceId());
+      assertThat(loggedSpans).contains(traces.get(i).get(0).getSpanId());
+    }
```

Comment:
I think it would be better to do these assertions inside the ` Consumer<TraceAssert> check` that would be more similar to what the groovy test did